libs/zedUpload/httputil/http.go

```diff
@@ -148,30 +150,112 @@ func ExecCmd(cmd, host, remoteFile, localFile string, objSize int64,
 			return stats
 		}
 		defer local.Close()
-		defer resp.Body.Close()
-		chunkSize := SingleMB
-		var written, copiedSize int64
-		stats.Size = objSize
-		for {
-			var copyErr error
-			if written, copyErr = io.CopyN(local, resp.Body, chunkSize); copyErr != nil && copyErr != io.EOF {
-				stats.Error = copyErr
-				return stats
+
+		var errorList []string
+		done := false
+		canSeek := false
+
+		delay := time.Second
+		lastModified := ""
+		var currentErrorList []string
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			if len(currentErrorList) > 0 {
+				// if we have previous errors, collect them
+				errorList = append(errorList, fmt.Sprintf("(attempt %d/%d):", attempt, maxRetries))
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				currentErrorList = []string{}
 			}
-			copiedSize += written
-			if written != chunkSize {
-				// Must have reached EOF
+			if ctx.Err() != nil {
+				errorList = append(errorList, fmt.Sprintf("ctx.Err: %v", ctx.Err()))
 				break
 			}
-			stats.Asize = copiedSize
-			if prgNotify != nil {
-				select {
-				case prgNotify <- stats:
-				default: //ignore we cannot write
+			if done {
+				break
+			}
+			if attempt > 0 {
+				time.Sleep(delay)
+				if delay < maxDelay {
+					delay = delay * 2
+				}
+			}
+			if canSeek && copiedSize > 0 {
+				req.Header.Set("Range", fmt.Sprintf("bytes=%d-", copiedSize))
+			}
+			resp, err := client.Do(req)
+			if err != nil {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("get failed: %s", err))
+				continue
+			}
+			if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+				currentErrorList = append(currentErrorList, fmt.Sprintf("bad response code: %d", resp.StatusCode))
+				err = resp.Body.Close()
+				if err != nil {
+					currentErrorList = append(currentErrorList, fmt.Sprintf("error close Body: %v", err))
+				}
+				errorList = append(errorList, strings.Join(currentErrorList, " "))
+				//we do not want to process server errors here
+				break
+			}
+			newLastModified := resp.Header.Get("Last-Modified")
+			if lastModified != "" && newLastModified != lastModified {
+				// last modified changed, retry without Range
+				lastModified = newLastModified
+				canSeek = false
+				continue
+			}
+			if resp.StatusCode == http.StatusOK {
+				stats.BodyLength = int(resp.ContentLength)
+			}
+			// canSeek affects next retry
+			canSeek = resp.Header.Get("Accept-Ranges") == "bytes"
+			if copiedSize > 0 && (!canSeek || resp.StatusCode != http.StatusPartialContent) {
```

Comment:
Why would we get a PartialContent http error? Wouldn't we just get an error that the connection failed?
Or does the http client go package insert the PartialContent status code?

Reply:
It is not an error, it is success code in response to the request with Range in header when server can process it.

Reply:
I don't understand which cases this if statement is supposed to detect.
One case is that Accept-Ranges was set on a previous response, but not on the current. 
I guess the server would also be broken if it didn't return StatusPartialContent.
But it seems like in both of those cases we shouldn't try with partial content again, because the server is misbehaving?

Reply:
It should detect cases when server do not response with "Accept-Ranges" or do not respond with StatusPartialContent (two variants: we do not put Range in headers or server do not respond with partial somehow). In both cases we will truncate file and start from beginning.